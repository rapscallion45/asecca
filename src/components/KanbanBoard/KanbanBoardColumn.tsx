import React, { FC } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Box, Typography } from '@mui/material';
import CircleIcon from '@mui/icons-material/Circle';
import { dragTask } from '@/redux/slices/kanbanSlice';
import { IKanbanBoardColumn, IKanbanBoardTask } from '@/lib/api/api-types';
import { AppState } from '@/redux/store';
import Task from './KanbanBoardTask';

/**
 * Kanban Board Column Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IKanbanBoardColumnProps
 * @prop {number} colIndex - column index
 */
interface IKanbanBoardColumnProps {
  colIndex: number;
}

/**
 * Kanban Board Column
 *
 * Kanban board column interface for displaying tasks in a certain category
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @component
 * @param {IKanbanBoardColumnProps} props - component props
 * @returns {FC} - data table functional component
 */
const KanbanBoardColumn: FC<IKanbanBoardColumnProps> = (props) => {
  const { colIndex } = props;
  const dispatch = useDispatch();
  const { data: kanbanData } = useSelector((state: AppState) => state.kanban);
  const colColors = ['secondary', 'warning', 'error', 'info', 'primary'];

  /* find this column in board state data from passed column index */
  const column = kanbanData.boards
    .find((board) => board.isActive === true)
    ?.columns.find(
      (col: IKanbanBoardColumn, index: number) => index === colIndex
    );

  /**
   * Callback listens for drop of task into column after drag
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.1
   *
   * @method
   * @param {any} event - object change event
   */
  const handleOnDrop = (event: any) => {
    const { prevColIndex, taskIndex } = JSON.parse(
      event.dataTransfer.getData('text')
    );

    if (colIndex !== prevColIndex) {
      dispatch(dragTask({ colIndex, prevColIndex, taskIndex }));
    }
  };

  /**
   * Callback listens for task drag over column
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.1
   *
   * @method
   * @param {any} event - object change event
   */
  const handleOnDragOver = (event: any) => {
    event.preventDefault();
  };

  /* if no column found in state datam return null */
  return column ? (
    <Box onDrop={handleOnDrop} onDragOver={handleOnDragOver} mr={2}>
      <Box display="flex" flexDirection="row" my={2}>
        {/* @ts-ignore */}
        <CircleIcon fontSize="small" color={colColors[colIndex]} />
        <Typography sx={{ ml: 1 }}>
          {column.name} ({column.tasks.length})
        </Typography>
      </Box>

      {column.tasks.map((task: IKanbanBoardTask, index: number) => (
        <Task key={task.title} taskIndex={index} colIndex={colIndex} />
      ))}
    </Box>
  ) : null;
};

export default KanbanBoardColumn;
