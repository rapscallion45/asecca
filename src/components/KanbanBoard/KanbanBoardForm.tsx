import { FC, useState, useCallback } from 'react';
import {
  Box,
  IconButton,
  Button,
  TextField,
  FormControl,
  Input,
  InputAdornment,
  CircularProgress,
  Typography,
} from '@mui/material';
import CloseIcon from '@mui/icons-material/Close';
import LabelImportantIcon from '@mui/icons-material/LabelImportant';
import { IKanbanBoardColumn } from '@/lib/api/api-types';
import { IKanbanBoardState } from '@/redux/types';
import useKanbanBoardFormController from './KanbanBoardFormController';
import { useSliceSelector } from '../SliceProvider/SliceProvider';

/**
 * Kanban Board Form Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.3
 *
 * @typedef IKanbanBoardFormProps
 * @prop {boolean} isEditMode - determines whether this is a new board or editing
 * @prop {boolean} canEdit - board data can be edited
 * @prop {any} closeModal - on close modal callback handler
 */
interface IKanbanBoardFormProps {
  isEditMode: boolean;
  canEdit?: boolean;
  closeModal?: () => void;
}

/**
 * Kanban Board Form
 *
 * Kanban board form interface for collecting user input when creating
 * or editing a board's details
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.3
 *
 * @component
 * @param {IKanbanBoardFormProps} props - component props
 * @returns {FC} - kanaban board form functional component
 */
const KanbanBoardForm: FC<IKanbanBoardFormProps> = (props) => {
  const { isEditMode, canEdit = false, closeModal } = props;
  const { data: currentData } = useSliceSelector() as IKanbanBoardState;
  const [newColumns, setNewColumns] = useState<Array<IKanbanBoardColumn>>(
    currentData?.columns || [
      { name: 'Todo', tasks: [] },
      { name: 'In Progress', tasks: [] },
      { name: 'Completed', tasks: [] },
    ]
  );
  const { saving, formik } = useKanbanBoardFormController(
    isEditMode,
    newColumns,
    closeModal
  );

  /**
   * Callback handler for user input updates to columns
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.3
   *
   * @method
   * @param {string} colIdx - column index that has been updated
   * @param {string} newValue - updated value for column name
   */
  const onChange = useCallback((colIdx: number, newValue: string) => {
    setNewColumns((prevState: Array<IKanbanBoardColumn>) => {
      const newState = [...prevState];
      const column = newState.find(
        (col: IKanbanBoardColumn, index: number) => index === colIdx
      );
      if (column) column.name = newValue;
      return newState;
    });
  }, []);

  /**
   * Callback handler for column deletion
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.3
   *
   * @method
   * @param {number} id - column ID to be deleted
   */
  const onDelete = useCallback((colIdx: number) => {
    setNewColumns((prevState) =>
      prevState.filter(
        (el: IKanbanBoardColumn, index: number) => index !== colIdx
      )
    );
  }, []);

  return (
    <form onSubmit={formik.handleSubmit}>
      <TextField
        fullWidth
        variant="outlined"
        margin="normal"
        id="name"
        name="name"
        label="Board Name"
        type="text"
        disabled={!canEdit}
        value={formik.values.name}
        onChange={formik.handleChange}
        error={formik.touched.name && Boolean(formik.errors.name)}
        helperText={formik.touched.name && formik.errors.name}
        autoComplete="on"
        InputProps={{
          startAdornment: (
            <InputAdornment position="start">
              <LabelImportantIcon color="secondary" />
            </InputAdornment>
          ),
        }}
      />
      <Box mt={2}>
        <Typography>Board Columns</Typography>
        {newColumns.map((column: IKanbanBoardColumn, index: number) => (
          <Box key={column.name}>
            <FormControl fullWidth sx={{ m: 1 }} variant="standard">
              <Input
                id="board-columns-form"
                type="text"
                value={column.name}
                onChange={(e) => {
                  onChange(index, e.target.value);
                }}
                disabled={!canEdit}
                placeholder="Enter column name..."
                endAdornment={
                  <InputAdornment position="end">
                    {canEdit && (
                      <IconButton
                        aria-label="delete column"
                        data-testid={`delete-column-${column.name}`}
                        onClick={() => {
                          onDelete(index);
                        }}
                      >
                        <CloseIcon />
                      </IconButton>
                    )}
                  </InputAdornment>
                }
              />
            </FormControl>
          </Box>
        ))}
        <Box mt={2}>
          <Button
            variant="outlined"
            color="secondary"
            onClick={() => {
              setNewColumns((state: Array<IKanbanBoardColumn>) => [
                ...state,
                { name: '', tasks: [], groups: [] },
              ]);
            }}
            fullWidth
            disabled={!canEdit}
          >
            + Add New Column
          </Button>
        </Box>
      </Box>
      {!isEditMode ? (
        <Button
          type="submit"
          fullWidth
          variant="contained"
          color="secondary"
          disabled={saving || !canEdit}
          sx={{ padding: '10px 0', marginTop: '20px' }}
        >
          {!saving && 'Add Board'}
          {saving && <CircularProgress size={25} color="inherit" />}
        </Button>
      ) : (
        <Button
          type="submit"
          fullWidth
          variant="contained"
          color="secondary"
          disabled={saving || !canEdit}
          sx={{ padding: '10px 0', marginTop: '20px' }}
        >
          {!saving && 'Update'}
          {saving && <CircularProgress size={25} color="inherit" />}
        </Button>
      )}
    </form>
  );
};

export default KanbanBoardForm;
