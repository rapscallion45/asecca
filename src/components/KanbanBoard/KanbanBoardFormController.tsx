import { useFormik } from 'formik';
import * as Yup from 'yup';
import { IKanbanBoardState } from '@/redux/types';
import { IKanbanBoard, IKanbanBoardColumn } from './types';
import { useSliceSelector } from '../SliceProvider/SliceProvider';

/**
 * Kanban board form controller hook, used for board form logic,
 * input handling and form submission
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.3
 * @memberof KanbanBoardForm
 *
 * @function
 * @param {boolean} isEditMode - is form creating a new board or editing existing task
 * @param {IKanbanBoard} currentData - board data
 * @param {Array<IKanbanBoardColumn>} newColumns - form column state data
 * @param {any} closeModal - callback for closing the board form modal
 */
const useKanbanBoardFormController = (
  isEditMode: boolean,
  newColumns: Array<IKanbanBoardColumn>,
  currentData?: IKanbanBoard,
  closeModal?: () => void
) => {
  const { saving } = useSliceSelector() as IKanbanBoardState;

  /**
   * Yup input validation configuration for board form
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.3
   *
   * @constant
   */
  const validationSchema = Yup.object().shape({
    name: Yup.string()
      .required('Board name is required')
      .max(60, 'Board name must not exceed 60 characters'),
  });

  /**
   * Formik configuration for board form
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.3
   *
   * @constant
   */
  const formik = useFormik({
    initialValues: {
      name: currentData?.name || 'New Board',
      newColumns: newColumns || [],
    },
    validationSchema,
    onSubmit: () => {
      if (closeModal && isEditMode) closeModal();
    },
  });

  return { saving, formik };
};
export default useKanbanBoardFormController;
