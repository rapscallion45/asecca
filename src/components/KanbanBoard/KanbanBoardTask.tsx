import React, { FC } from 'react';
import { useSelector } from 'react-redux';
import { Card, CardContent, Typography } from '@mui/material';
import {
  IKanbanBoard,
  IKanbanBoardColumn,
  IKanbanBoardTask,
} from '@/lib/api/api-types';
import { AppState } from '@/redux/store';
// import TaskModal from "../modals/TaskModal";

/**
 * Kanban Board Task Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @typedef IKanbanBoardTaskProps
 * @prop {number} colIndex - column index that task is in
 * @prop {number} taskIndex - index of task
 */
interface IKanbanBoardTaskProps {
  colIndex: number;
  taskIndex: number;
}

/**
 * Kanban Board Task
 *
 * Kanban board task interface for displaying task overview from board column
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @component
 * @param {IKanbanBoardTaskProps} props - component props
 * @returns {FC} - data table functional component
 */
const KanbanBoardTask: FC<IKanbanBoardTaskProps> = (props) => {
  const { colIndex, taskIndex } = props;
  // const [isTaskModalOpen, setIsTaskModalOpen] = useState<boolean>(false);
  const { data: kanbanData } = useSelector((state: AppState) => state.kanban);

  /* find this task and subtasks from passed board column and task indicies */
  const task = kanbanData.boards
    ?.find((board: IKanbanBoard) => board.isActive === true)
    ?.columns.find(
      (col: IKanbanBoardColumn, index: number) => index === colIndex
    )
    ?.tasks.find(
      (taskItem: IKanbanBoardTask, index: number) => index === taskIndex
    );
  const subtasks = task?.subtasks;

  /* calculate completed subtask number for this task */
  let completed = 0;
  subtasks?.forEach((subtask) => {
    if (subtask.isCompleted) {
      completed += 1;
    }
  });

  /**
   * Callback listens for drag of task
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.1
   *
   * @method
   * @param {any} event - object change event
   */
  const handleOnDrag = (event: any) => {
    event.dataTransfer.setData(
      'text',
      JSON.stringify({ taskIndex, prevColIndex: colIndex })
    );
  };

  return task ? (
    <>
      {/* <div
        onClick={() => {
          setIsTaskModalOpen(true);
        }}
        draggable
        onDragStart={handleOnDrag}
        className=" w-[280px] first:my-5 rounded-lg  bg-white  dark:bg-[#2b2c37] shadow-[#364e7e1a] py-6 px-3 shadow-lg hover:text-[#635fc7] dark:text-white dark:hover:text-[#635fc7] cursor-pointer "
      >
        <p className=" font-bold tracking-wide ">{task?.title}</p>
        <p className=" font-bold text-xs tracking-tighter mt-2 text-gray-500">
          {completed} of {subtasks?.length} completed tasks
        </p>
      </div> */}
      <Card
        draggable
        // onClick={() => {
        //   setIsTaskModalOpen(true);
        // }}
        onDragStart={handleOnDrag}
        sx={{ width: 275 }}
      >
        <CardContent>
          <Typography sx={{ fontSize: 14 }} color="text.secondary" gutterBottom>
            Task
          </Typography>
          <Typography variant="h5" component="div">
            {task?.title}
          </Typography>
          <Typography sx={{ mb: 1.5 }} color="text.secondary">
            {completed} of {subtasks?.length} tasks complete
          </Typography>
          <Typography variant="body2">{task?.description}</Typography>
        </CardContent>
      </Card>
      {/* {isTaskModalOpen && (
        <TaskModal
          colIndex={colIndex}
          taskIndex={taskIndex}
          setIsTaskModalOpen={setIsTaskModalOpen}
        />
      )} */}
    </>
  ) : null;
};

export default KanbanBoardTask;
