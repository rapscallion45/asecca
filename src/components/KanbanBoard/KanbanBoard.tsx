import { FC } from 'react';
import {
  Box,
  Skeleton,
  Button,
  Card,
  CardContent,
  useTheme,
} from '@mui/material';
import { LoadingButton } from '@mui/lab';
import ViewColumnIcon from '@mui/icons-material/ViewColumn';
import KanbanBoardEmpty from '@/components/KanbanBoard/KanbanBoardEmpty/KanbanBoardEmpty';
import Column from '@/components/KanbanBoard/KanbanBoardColumn/KanbanBoardColumn';
import ScrollBar from '@/components/ScrollBar/ScrollBar';
import FormModal from '@/modals/FormModal/FormModal';
import { IKanbanBoardState } from '@/redux/types';
import KanbanBoardForm from './KanbanBoardForm';
import { IKanbanBoard, IKanbanBoardColumn } from './types';
import { useSliceSelector } from '../SliceProvider/SliceProvider';

/**
 * Kanban Board Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.7
 *
 * @typedef IKanbanBoardProps
 * @prop {IKanbanBoard} currentData - passed board data
 */
interface IKanbanBoardProps {
  currentData: IKanbanBoard;
}

/**
 * Kanban Board
 *
 * Presents columns and task data for passed kanban board
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.7
 *
 * @component
 * @returns {FC} - Kanban Board interface functional component
 */
const KanbanBoard: FC<IKanbanBoardProps> = (props) => {
  const { currentData } = props;
  const theme = useTheme();
  //   const [isBoardModalOpen, setIsBoardModalOpen] = useState(false);
  const { loading, error, saving, edited } =
    useSliceSelector() as IKanbanBoardState;

  /**
   * Callback handler for cancelling changes to board
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.7
   *
   * @method
   */
  const handleCancel = () => {};

  /**
   * Callback handler for saving changes to board
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.7
   *
   * @method
   */
  const handleSave = () => {};

  /**
   * Helper component for rendering column skeleton whilst loading board
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.7
   *
   * @method
   */
  const renderBoardColumnSkeleton = () => (
    <Box mr={2}>
      <Skeleton width={190} height={50} sx={{ borderRadius: '4px', mb: 1 }} />
      <Skeleton
        variant="rectangular"
        width={270}
        height={170}
        sx={{ borderRadius: '4px', mb: 2 }}
      />
      <Skeleton
        variant="rectangular"
        width={270}
        height={170}
        sx={{ borderRadius: '4px', mb: 2 }}
      />
    </Box>
  );

  return (
    <>
      <ScrollBar>
        {currentData.columns.length > 0 ? (
          <Box display="flex" flexDirection="row" sx={{ pt: 2, pb: 1 }}>
            {!loading && !error ? (
              <>
                {currentData.columns.map((col: IKanbanBoardColumn, index) => (
                  <Column
                    key={col.id}
                    colIndex={index}
                    columns={currentData.columns}
                  />
                ))}
                <Card
                  sx={{
                    mt: 7,
                    minWidth: 275,
                    minHeight: 300,
                    maxHeight: 500,
                    backgroundColor: 'transparent',
                    borderColor: theme.palette.text.secondary,
                    border: 'solid 1px',
                  }}
                >
                  <CardContent sx={{ padding: '50% 40px' }}>
                    <FormModal
                      triggerBtn={{
                        type: 'menu',
                        // @ts-ignore
                        icon: ViewColumnIcon,
                        iconStyle: { marginRight: '10px' },
                        text: '+ Add Column',
                        closeMenu: () => {},
                      }}
                      title="Add Column"
                    >
                      <KanbanBoardForm
                        isEditMode
                        currentData={currentData}
                        closeModal={() => {}}
                      />
                    </FormModal>
                  </CardContent>
                </Card>
              </>
            ) : (
              <>
                {renderBoardColumnSkeleton()}
                {renderBoardColumnSkeleton()}
                {renderBoardColumnSkeleton()}
              </>
            )}
          </Box>
        ) : (
          <Box mt={2}>
            <KanbanBoardEmpty type="edit" currentData={currentData} />
          </Box>
        )}
      </ScrollBar>
      <Box
        sx={{
          marginTop: 4,
          display: 'flex',
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'center',
        }}
      >
        <Button
          color="secondary"
          variant="outlined"
          onClick={handleCancel}
          disabled={saving || loading || !edited}
          sx={{ backgroundColor: 'common.white' }}
        >
          Reset Board
        </Button>
        <LoadingButton
          color="secondary"
          variant="contained"
          onClick={handleSave}
          disabled={saving || loading || !edited}
          loading={saving}
          sx={{ ml: 2 }}
        >
          Save
        </LoadingButton>
      </Box>
    </>
  );
};

export default KanbanBoard;
