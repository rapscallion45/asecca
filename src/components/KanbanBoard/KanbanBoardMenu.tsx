import { FC, useState, useCallback, MouseEvent } from 'react';
import EditOutlinedIcon from '@mui/icons-material/EditOutlined';
import MoreVertIcon from '@mui/icons-material/MoreVert';
import { alpha } from '@mui/material/styles';
import { Button, Box, IconButton } from '@mui/material';
// import { AppState } from '@/redux/store';
import { IKanbanBoard } from '@/lib/api/api-types';
import MenuPopover from '@/components/MenuPopover/MenuPopover';
import FormModal from '@/modals/FormModal/FormModal';
import { ModalButtonIconSizeType } from '@/modals/types';
import KanbanBoardForm from './KanbanBoardForm';

/**
 * Kanban Board Menu Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.3
 *
 * @typedef IKanbanBoardMenuProps
 * @prop {IKanbanBoard} currentData - current board data
 * @prop {ModalButtonIconSizeType} iconSize - button icon size
 */
interface IKanbanBoardMenuProps {
  currentData: IKanbanBoard;
  iconSize?: ModalButtonIconSizeType;
}

/**
 * Kanban Board Menu
 *
 * Kanbaord board menu interface
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.3
 *
 * @component
 * @param {IKanbanBoardMenuProps} props - component props
 * @returns {FC} - kanban board menu functional component
 */
const KanbanBoardkMenu: FC<IKanbanBoardMenuProps> = (props) => {
  const { currentData, iconSize } = props;
  // const { deleting } = useSelector((state: AppState) => state.bugs);
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);

  /**
   * Callback listener for opening board menu popover
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.2
   *
   * @method
   * @param {MouseEvent<HTMLButtonElement>} event - object change event
   */
  const handleOpenMenu = useCallback((event: MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  }, []);

  /**
   * Callback listener for closing board menu popover
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.2
   *
   * @method
   */
  const handleCloseMenu = useCallback(() => {
    setAnchorEl(null);
  }, []);

  return (
    <>
      <IconButton
        size="small"
        color={anchorEl ? 'primary' : 'default'}
        onClick={handleOpenMenu}
        sx={{
          ...(Boolean(anchorEl) && {
            bgcolor: (theme) =>
              alpha(
                theme.palette.primary.main,
                theme.palette.action.focusOpacity
              ),
          }),
        }}
      >
        <MoreVertIcon fontSize={iconSize || 'medium'} />
      </IconButton>
      <MenuPopover
        open={Boolean(anchorEl)}
        onClose={handleCloseMenu}
        keepMounted
        anchorEl={anchorEl}
        sx={{ width: 220 }}
      >
        <FormModal
          triggerBtn={{
            type: 'menu',
            // @ts-ignore
            icon: EditOutlinedIcon,
            iconStyle: { marginRight: '10px' },
            text: 'Edit Board Info',
            closeMenu: handleCloseMenu,
          }}
          title="Edit Board Info"
        >
          <KanbanBoardForm
            isEditMode
            currentData={currentData}
            closeModal={handleCloseMenu}
          />
        </FormModal>
        <Box sx={{ p: 2, pt: 1.5 }}>
          <Button
            fullWidth
            color="inherit"
            variant="outlined"
            onClick={handleCloseMenu}
          >
            Close
          </Button>
        </Box>
      </MenuPopover>
    </>
  );
};
export default KanbanBoardkMenu;
