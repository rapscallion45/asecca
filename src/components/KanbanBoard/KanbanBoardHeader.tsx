import { FC } from 'react';
import {
  Box,
  Typography,
  Skeleton,
  Divider,
  LinearProgress,
} from '@mui/material';
import KanbanBoardMenu from '@/components/KanbanBoard/KanbanBoardMenu';
import { useSliceSelector } from '@/components/SliceProvider/SliceProvider';
import { IKanbanBoardState } from '@/redux/types';

/**
 * Kanban Board Header Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.11
 *
 * @typedef IKanbanBoardHeaderProps
 * @prop {string} name - Kanban board name
 */
interface IKanbanBoardHeaderProps {
  name: string;
}

/**
 * Kanban Board Header
 *
 * Application Kanban Board Header interface
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.9
 *
 * @component
 * @returns {FC} - Kanban Board Header functional component
 */
const KanbanBoardHeader: FC<IKanbanBoardHeaderProps> = (props) => {
  const { name } = props;
  const {
    data: boardData,
    error,
    loading,
  } = useSliceSelector() as IKanbanBoardState;

  return (
    <>
      <Box display="flex" pb={1}>
        <Typography variant="h4">
          {!loading && !error ? (
            name
          ) : (
            <Box display="flex">
              <Typography variant="h4">
                {error ?? 'Loading board...'}
              </Typography>
            </Box>
          )}
        </Typography>
        {!loading && !error ? (
          <Box display="flex" justifyContent="end" sx={{ flexGrow: 1 }}>
            <Box display="flex" justifyContent="center" alignItems="center">
              <KanbanBoardMenu currentData={boardData} />
            </Box>
          </Box>
        ) : (
          <Box display="flex" justifyContent="end" sx={{ flexGrow: 1 }}>
            {!error && (
              <>
                <Skeleton
                  variant="rectangular"
                  width={140}
                  height={40}
                  sx={{ borderRadius: '4px' }}
                />
                <Skeleton
                  variant="rectangular"
                  width={70}
                  height={40}
                  sx={{ borderRadius: '4px', ml: 2 }}
                />
              </>
            )}
          </Box>
        )}
      </Box>
      {!loading ? (
        <Divider />
      ) : (
        <LinearProgress color="secondary" sx={{ borderRadius: 1 }} />
      )}
    </>
  );
};

export default KanbanBoardHeader;
