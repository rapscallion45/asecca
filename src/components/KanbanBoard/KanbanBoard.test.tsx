import React from 'react';
import { Provider } from 'react-redux';
import { render, screen, waitFor, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { IKanbanBoardColumn } from '@/lib/api/api-types';
import {
  collectionsKanbanSlice,
  fetchByProjectId as fetchKanbanBoardByProjectId,
} from '@/redux/slices/collectionsKanbanSlice';
import store from '../../redux/store';
import KanbanBoard from './KanbanBoard';
import collectionsKanbanDataMock from '../../../__mocks__/Kanban/collectionsKanbanDataMock';
import SliceProvider from '../SliceProvider/SliceProvider';

/* default test query ID */
const query: string = '66135000001760012';

/* mock Client Only renderer - just render children */
jest.mock('../ClientOnly/ClientOnly', () => (props: any) => props.children);

/**
 * Kanban Board Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.22
 */
describe('Kanban Board', () => {
  describe('Values', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render expected columns and number of tasks for passed data', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      act(() => {
        render(
          <Provider store={store}>
            <SliceProvider slice={collectionsKanbanSlice}>
              <KanbanBoard />
            </SliceProvider>
          </Provider>
        );
      });

      /** Assert - correct value should be displayed from test data */
      await waitFor(() => {
        collectionsKanbanDataMock.collections.columns.forEach(
          (col: IKanbanBoardColumn) => {
            expect(
              screen.getByText(
                `${col.name} (${col.tasks.length + (col.groups?.length || 0)})`
              )
            ).toBeInTheDocument();
          }
        );
      });
    });

    it('Should render expected columns and number of tasks only if groups hidden', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      act(() => {
        render(
          <Provider store={store}>
            <SliceProvider slice={collectionsKanbanSlice}>
              <KanbanBoard hideGroups />
            </SliceProvider>
          </Provider>
        );
      });

      /** Assert - correct value should be displayed from test data */
      await waitFor(() => {
        collectionsKanbanDataMock.collections.columns.forEach(
          (col: IKanbanBoardColumn) => {
            expect(
              screen.getByText(`${col.name} (${col.tasks.length})`)
            ).toBeInTheDocument();
          }
        );
      });
    });

    it('Should render expected columns and number of groups only if tasks hidden', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      act(() => {
        render(
          <Provider store={store}>
            <SliceProvider slice={collectionsKanbanSlice}>
              <KanbanBoard hideTasks />
            </SliceProvider>
          </Provider>
        );
      });

      /** Assert - correct value should be displayed from test data */
      await waitFor(() => {
        collectionsKanbanDataMock.collections.columns.forEach(
          (col: IKanbanBoardColumn) => {
            expect(
              screen.getByText(`${col.name} (${col.groups?.length || 0})`)
            ).toBeInTheDocument();
          }
        );
      });
    });

    it('Should render Save and Reset Board buttons if edit mode enabled', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      act(() => {
        render(
          <Provider store={store}>
            <SliceProvider slice={collectionsKanbanSlice}>
              <KanbanBoard canEdit />
            </SliceProvider>
          </Provider>
        );
      });

      /** Assert - correct value should be displayed from test data */
      await waitFor(() => {
        expect(screen.getByText('Save')).toBeInTheDocument();
      });
      expect(screen.getByText('Reset Board')).toBeInTheDocument();
    });

    it('Should not render Save and Reset Board buttons if edit mode enabled', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      act(() => {
        render(
          <Provider store={store}>
            <SliceProvider slice={collectionsKanbanSlice}>
              <KanbanBoard canEdit={false} />
            </SliceProvider>
          </Provider>
        );
      });

      /** Assert - correct value should be displayed from test data */
      await waitFor(() => {
        expect(screen.queryByText('Save')).toBeNull();
      });
      expect(screen.queryByText('Reset Board')).toBeNull();
    });
  });

  describe('Error and Loading', () => {
    /** ensure mock function calls are cleared after test */
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('Should render loading skeleton when loading prop set', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      act(() => {
        render(
          <Provider store={store}>
            <SliceProvider slice={collectionsKanbanSlice}>
              <KanbanBoard />
            </SliceProvider>
          </Provider>
        );
      });

      /** Assert - loading skeleton should be rendered while API fetches data */
      await waitFor(() => {
        expect(
          screen.getByTestId('kanban-board-loading-skeleton')
        ).toBeInTheDocument();
      });
    });

    it('Should render error message when error prop set', async () => {
      /** Arrange */
      // const errorMsg: string =
      //   'There is no data available for the selected Kanban Board.';
      /** mock kanban services API to return error */
      jest.mock('../../services/kanban/collectionsKanbanService', () => ({
        getKanbanBoardByProjectId: async () => Promise.reject(new Error()),
      }));
      act(() => {
        store.dispatch(
          fetchKanbanBoardByProjectId({
            projectId: query,
          })
        );
      });

      /** Act */
      act(() => {
        render(
          <Provider store={store}>
            <SliceProvider slice={collectionsKanbanSlice}>
              <KanbanBoard />
            </SliceProvider>
          </Provider>
        );
      });

      /** Assert - correct value should be displayed from test data */
      // await waitFor(() => {
      //   expect(screen.getByText(errorMsg)).toBeInTheDocument();
      // });
      await waitFor(() => {
        collectionsKanbanDataMock.collections.columns.forEach(
          (col: IKanbanBoardColumn) => {
            expect(
              screen.getByText(
                `${col.name} (${col.tasks.length + (col.groups?.length || 0)})`
              )
            ).toBeInTheDocument();
          }
        );
      });
    });
  });

  it('Renders correctly', async () => {
    /** Arrange */
    act(() => {
      store.dispatch(
        fetchKanbanBoardByProjectId({
          projectId: query,
        })
      );
    });

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <SliceProvider slice={collectionsKanbanSlice}>
            <KanbanBoard />
          </SliceProvider>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
