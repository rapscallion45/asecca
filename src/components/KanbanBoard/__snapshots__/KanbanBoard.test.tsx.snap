// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kanban Board Renders correctly 1`] = `
<div
  data-simplebar="init"
  onMouseDown={[Function]}
>
  <div
    className="simplebar-wrapper"
  >
    <div
      className="simplebar-height-auto-observer-wrapper"
    >
      <div
        className="simplebar-height-auto-observer"
      />
    </div>
    <div
      className="simplebar-mask"
    >
      <div
        className="simplebar-offset"
      >
        <div
          aria-label="scrollable content"
          className="simplebar-content-wrapper"
          role="region"
          tabIndex={0}
        >
          <div
            className="simplebar-content"
          >
            <div
              className="MuiBox-root css-8npigc"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="simplebar-placeholder"
    />
  </div>
  <div
    className="simplebar-track simplebar-horizontal"
  >
    <div
      className="simplebar-scrollbar"
    />
  </div>
  <div
    className="simplebar-track simplebar-vertical"
  >
    <div
      className="simplebar-scrollbar"
    />
  </div>
</div>
`;
